hera_cal/abscal.py

```diff
@@ -982,113 +988,129 @@ def abscal_run(data_files, model_files, verbose=True, overwrite=False, write_cal
-                echo("it is recommended to run a delay_cal before avg_phs_cal", verbose=verbose)
-            if all_antenna_gains:
-                raise ValueError("can't run avg_phs_cal when all_antenna_gains is True")
-            AC.phs_logcal(avg=True, verbose=verbose)
-            AC.data = apply_gains(AC.data, AC.ant_phi_gain, gain_convention='divide')
-            gain_list.append(AC.ant_phi_gain)
-
-        if delay_slope_cal:
-            AC.delay_slope_lincal(verbose=verbose, time_avg=True)
-            AC.data = apply_gains(AC.data, AC.dly_slope_gain, gain_convention='divide')
-            if all_antenna_gains:
-                gain_list.append(AC.custom_dly_slope_gain(total_gain_keys, total_data_antpos))
-            else:
-                gain_list.append(AC.dly_slope_gain)
-
-        if abs_amp_cal:
-            AC.abs_amp_logcal(verbose=verbose)
-            AC.data = apply_gains(AC.data, AC.abs_eta_gain, gain_convention='divide')
-            if all_antenna_gains:
-                gain_list.append(AC.custom_abs_eta_gain(total_gain_keys))
-            else:
-                gain_list.append(AC.abs_eta_gain)
-
-        if TT_phs_cal:
-            if delay_slope_cal == False:
-                echo("it is recommended to run a delay_slope_cal before TT_phs_cal", verbose=verbose)
-            AC.TT_phs_logcal(verbose=verbose)
-            AC.data = apply_gains(AC.data, AC.TT_Phi_gain)
-            if all_antenna_gains:
-                gain_list.append(AC.custom_TT_Phi_gain(total_gain_keys, total_data_antpos))
-                gain_list.append(AC.custom_abs_psi_gain(total_gain_keys))
-            else:
-                gain_list.append(AC.abs_psi_gain)
-                gain_list.append(AC.TT_Phi_gain)
-
-        if gen_amp_cal:
-            if all_antenna_gains:
-                raise ValueError("can't run gen_amp_cal when all_antenna_gains is True")
-            AC.amp_logcal(verbose=verbose)
-            AC.data = apply_gains(AC.data, AC.ant_eta_gain, gain_convention='divide')
-            gain_list.append(AC.ant_eta_gain)
-
-        if gen_phs_cal:
-            if delay_cal == False and delay_slope_cal == False:
-                echo("it is recommended to run a delay_cal or delay_slope_cal before gen_phs_cal", verbose=verbose)
-            if all_antenna_gains:
-                raise ValueError("can't run gen_phs_cal when all_antenna_gains is True")
-            AC.phs_logcal(verbose=verbose)
-            AC.data = apply_gains(AC.data, AC.ant_phi_gain, gain_convention='divide')
-            gain_list.append(AC.ant_phi_gain)
-
-        # collate gains
-        if len(gain_list) == 0:
-            raise ValueError("abscal_run executed without any calibration arguments set to True")
-        gain_dict = merge_gains(gain_list)
+            flag_dict = odict(map(lambda k: (k, np.ones((Ntimes, Nfreqs), np.bool)), gain_dict.keys()))
+
+        # load calfits file if provided
+        if calfits_files is not None:
```

Comment:
I think a lot of this could be much more easily accomplished with `hera_cal.io.update_cal`.

Reply:
yeah i guess i was trying to limit I/O, so the idea was to this in memory before having to write out the calfits file...

Reply:
Calfits I/O is trivial and will never be an issue compared to visibility I/O.